### ISSUE TITLE
Proposed Change: When stubbing methods, use a block instead of and_return

### ISSUE THREAD

**Issue by joshmfrankel** [lessonly/lessonly.github.io#109]
Currently our styleguide recommends the following:

> When stubbing methods, use a block instead of and_return.#
> Because expect().and_return() sounds like you’re expecting the method to return something, when in fact you’re forcing it to. For example:
> 
> expect(UserUpdate).to receive(:new).with(anything) { user_update }

I've recently been working with `.and_yield( )` which functions similarly to `.and_return( )` except for a block. Given that both response setups have similar format for me it seems confusing to use the block syntax for `and_return` and the method syntax for `and_yield`. I believe that using the explicit `and_return` will increase readability and understanding.

My proposal is to amend the current styleguide issue that favors block syntax and instead prefer `and_return` method syntax.

**Comment by unixmonkey** [comment]
I just wanted to voice that I support allowing `and_return`. Where possible, I like my code and specs to (somewhat) make sense when read aloud, and the block syntax irks me a bit. I kind of understand the reasoning given, but I don't think it's a good enough reason to push the preference.

If anything, I think it's a bit weird to use `expect` instead of `allow` for most stubs, but I get it that having a built-in expectation on them can save you troubleshooting sometimes. Consistency is nice, but sometimes that comes at a cost of limiting expressiveness.

**Comment by rreinhardt9** [comment]
Great thought Josh, thanks for bringing this up! I wonder if we can just get rid of the block syntax requirement entirely? The block syntax reads a little weird to me as well. Additionally, it seems like our justification for requiring it is kind of weak:

> sounds like you’re expecting the method to return something

I'm not positive, but the only possibly practical difference that I could think of it that the block is probably lazy evaluated and the method would be evaluated eagerly? Not sure that would make a difference in most cases even if it was true in the tests?

An argument could be made to leave this up to the authors discretion so that they can write the most expressive test?

I personally don't really prefer either one over the other; but it's nice to have options. It seems like the broader community (at least examples I see in the wild) tend to use `and_return`? 

> Consistency is nice, but sometimes that comes at a cost of limiting expressiveness.

I totally agree there. This could be a case where we are being over prescriptive and could simplify to allow expressiveness?

**Comment by tmaconnelly** [comment]
FWIW, I prefer `and_return` as well, I can't remember where I read it, but I once read that your specs should read like documentation for your code. And so as Dave mentions, I think it's nice that it sounds more like an english sentence. 🤷‍♂️

**Comment by stevegrossi** [comment]
As (I think: de2152886ab1c3b694f9dcf5244f8d1d19250a32) the original author of the rule, I support this amendment! The main motivation was clarity, so if we're finding an alternative more clear, let's use it.

I've also wanted this for purely syntactic reasons, since it's nice to be able to chain long stubs like

```ruby
expect(Something)
  .to_receive(:some_method)
  .with(:a, :bunch, :of:, :arguments, oh: "Hi", this: "is", a: "Hash")
  .and_return(some_long_variable_name_pointing_to_a_value)
```

which the block syntax interferes with.

**Comment by armilam** [comment]
@stevegrossi I remember you and I discussed this way back and agreed the block was more clear. But I'm also a fan of doing things the way the community does them. And the `and_return` syntax is probably slightly faster than the block. And chaining is nice.

I support this proposal.

**Comment by wuz** [comment]
For those of us that don't write ruby tests often, could someone add a quick example of the block syntax? I agree with all the statements so far, but would love to see what we are talking about 😆

**Comment by tmaconnelly** [comment]
`expect(Class).to receive(:perform).and_return(true)`

vs
the block syntax:

`expect(Class).to receive(:perform) { true }`

**Comment by joshmfrankel** [comment]
I posted the changeset necessary to make this happen. Instead of requiring the .and_return syntax I categorized it as recommended / not recommended

**Comment by case-eee** [comment]
I also support this proposal. Thanks for taking the initiative to make this change 😄

**Comment by stevegrossi** [comment]
Resolved by #111 🎉

**Comment by stevegrossi** [comment]
Resolved by #111 🎉

**Comment by rreinhardt9** [comment]
I'm curious... why list a preference at all rather than leave it up to the developers discretion to determine what gives the most clarity? I guess it's not a problem to now list that we prefer `and_return` it just seems unnecessary.
